### PR TITLE
tests pass locally...

### PR DESCRIPTION
**Pull request by chriselrod**

### CODE REVIEW

**Comment by chriselrod** [comment]
Lol, okay, I know why it's segfaulting. Silly mistake; it'll only segfault if the script guesses you have a single physical core or 1 thread, which is the case on CI...

File: src/optimize.jl [hunk 1]
```diff
@@ -21,7 +21,7 @@ function update!(o::ADAM, (mt, vt, βp), x, Δ)
     vt[i] = β₂ * vt[i] + (1 - β₂) * Δ[i]^2
     # Δ[i] =  mt[i] / (1 - βp₁) / (sqrt(vt[i] / (1 - βp₂)) + 1e-8) * η
     Δᵢ = η * mt[i] / ((1 - βp₁) * (sqrt(vt[i] / (1 - βp₂)) + 1e-8))
-    Δ[i] = Δᵢ
+    # Δ[i] = Δᵢ
     x[i] -= Δᵢ#Δ[i]
```
Comment (korsbo) [inline src/optimize.jl:25]: Clean up?

File: src/penalty.jl [hunk 2]
```diff
@@ -84,20 +84,21 @@ L1Penalty(p::AbstractPenalty, λ) = L1Penalty(getchain(p), λ)
 getλ(p::L1Penalty) = getfield(p, :λ)
 (p::L1Penalty)(chn::SimpleChain) = L1Penalty(chn, p.λ)
 
-@inline function apply_penalty(Λ::L1Penalty{NN,T2}, p::AbstractVector{T3}) where {T2,T3,NN}
+@inline function apply_penalty(Λ::L1Penalty, p::AbstractVector{T3}) where {T3}
```
Comment (korsbo) [inline src/penalty.jl:87]: Nit: no need for `3` in `T3` anymore.

File: src/penalty.jl [hunk 3]
```diff
@@ -84,20 +84,21 @@ L1Penalty(p::AbstractPenalty, λ) = L1Penalty(getchain(p), λ)
...
     l += abs(p[i])
   end
-  Base.FastMath.mul_fast(l, Λ.λ)
+  Base.FastMath.mul_fast(l, T3(Λ.λ))
 end
 function apply_penalty!(
   g::AbstractVector{T1},
-  Λ::L1Penalty{NN,T2},
+  Λ::L1Penalty,
   p::AbstractVector{T3},
```
Comment (korsbo) [inline src/penalty.jl:97]: Nit: `T3` -> `T2`

File: src/penalty.jl [hunk 4]
```diff
@@ -118,21 +119,22 @@ L2Penalty(p::AbstractPenalty, λ) = L2Penalty(getchain(p), λ)
...
     l += pᵢ * pᵢ
   end
-  Base.FastMath.mul_fast(l, Λ.λ)
+  Base.FastMath.mul_fast(l, T(Λ.λ))
 end
 function apply_penalty!(
   g::AbstractVector{T1},
-  Λ::L2Penalty{NN,T2},
+  Λ::L2Penalty,
   p::AbstractVector{T3},
```
Comment (korsbo) [inline src/penalty.jl:133]: Nit: 3 -> 2

File: src/simple_chain.jl [hunk 5]
```diff
@@ -346,14 +349,17 @@ function chain_valgrad!(
...
   lgrad, pu4 = pullback!(pg, l, grad, arg, p, pu, pu3)
+  # @show X1,lgrad
   return val, lgrad, pu4
 end
 function chain_valgrad!(pg, arg, layers::Tuple{X}, p::Ptr, pu::Ptr{UInt8}) where {X}
   l = getfield(layers, 1)
   __, val, _, pu2 = valgrad_layer!(pg, l, arg, p, pu)
   # val, pullback, p2, pu2 = valgrad_layer!(pg, l, arg, p, pu)
   lgrad, pu3 = pullback!(pg, l, One(), arg, p, pu, pu2)
+  # @show lgrad
```
Comment (korsbo) [inline src/simple_chain.jl:362]:
```suggestion
```

File: src/simple_chain.jl [hunk 6]
```diff
@@ -346,14 +349,17 @@ function chain_valgrad!(
   l = getfield(layers, 1)
   pg2, larg, p2, pu2 = valgrad_layer!(pg, l, arg, p, pu)
   val, grad, pu3 = chain_valgrad!(pg2, larg, Base.tail(layers), p2, pu2)
+  # @show X2,grad
   lgrad, pu4 = pullback!(pg, l, grad, arg, p, pu, pu3)
+  # @show X1,lgrad
```
Comment (korsbo) [inline src/simple_chain.jl:354]:
```suggestion
  lgrad, pu4 = pullback!(pg, l, grad, arg, p, pu, pu3)
```

**Review by korsbo — COMMENTED** [review]
A few nit-picks and some added functions not covered by tests (see codecov). Otherwise fine.

**Comment by chriselrod** [comment]
> A few nit-picks and some added functions not covered by tests (see codecov). Otherwise fine.

Fixed. We should get this in for the 0.2.1 release.

**Review by korsbo — APPROVED** [review]

**Pull request merged**